### PR TITLE
Simplify inlining intermediate polynomials.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: ast/src/analyzed/mod.rs [hunk 1]
```diff
@@ -357,43 +356,38 @@ fn substitute_intermediate<T: Copy>(
...
+        if poly_to_replace.next && r.next {
+            let value = intermediate_polynomials[&poly_to_replace.poly_id];
+            panic!("Error inlining intermediate polynomial {poly_to_replace} = {value}:\nNext operator already applied to {poly_to_replace} but also in {r} - cannot apply it twice!");
+        }
+        r.next = poly_to_replace.next;
+        match r.poly_id.ptype {
+            PolynomialType::Committed | PolynomialType::Constant => {}
+            PolynomialType::Intermediate => {
+                *e = inlined_expression_from_intermediate_poly_id(
+                    (*r).clone(),
```
Comment (georgwiese) [inline ast/src/analyzed/mod.rs:383]:
```suggestion
                    r.clone(),
```
This should work too?

File: ast/src/analyzed/mod.rs [hunk 2]
```diff
@@ -357,43 +356,38 @@ fn substitute_intermediate<T: Copy>(
...
-                                cache,
-                                r.next,
-                            )
-                        });
-                }
+        let AlgebraicExpression::Reference(r) = e else { return; };
+        // "forward" the next operator from the polynomial to be replaced.
+        if poly_to_replace.next && r.next {
+            let value = intermediate_polynomials[&poly_to_replace.poly_id];
+            panic!("Error inlining intermediate polynomial {poly_to_replace} = {value}:\nNext operator already applied to {poly_to_replace} but also in {r} - cannot apply it twice!");
```
Comment (georgwiese) [inline ast/src/analyzed/mod.rs:376]:
```suggestion
            panic!("Error inlining intermediate polynomial {poly_to_replace} = {value}':\nNext operator already applied to {} and then again to {} - cannot apply it twice!", r.name, poly_to_replace.name);
```

A bit clearer in my opinion. Example:
```
Error inlining intermediate polynomial main_arith.carry' = ((main_arith.carry_high' * 65536) + main_arith.carry_low)':
Next operator already applied to main_arith.carry_high and then again to main_arith.carry - cannot apply it twice!
```

**Review by georgwiese — APPROVED** [review]

File: ast/src/analyzed/mod.rs (hunk 2 above)
Comment (chriseth) [reply ast/src/analyzed/mod.rs:376]: true!

File: ast/src/analyzed/mod.rs [hunk 3]
```diff
@@ -357,43 +356,42 @@ fn substitute_intermediate<T: Copy>(
...
+        // "forward" the next operator from the polynomial to be replaced.
+        if poly_to_replace.next && r.next {
+            let value = intermediate_polynomials[&poly_to_replace.poly_id];
+            panic!(
+                "Error inlining intermediate polynomial {poly_to_replace} = ({value})':\nNext operator already applied to {} and then again to {} - cannot apply it twice!",
+                r.name,
+                poly_to_replace.name
+            );
+        }
+        r.next = r.next || poly_to_replace.next;
```
Comment (chriseth) [inline ast/src/analyzed/mod.rs:382]: forgot the `r.next ||`

**Review by georgwiese — APPROVED** [review]
LGTM!

**Pull request merged**